common/numerated/src/tree.rs

```diff
@@ -0,0 +1,835 @@
+/// tree.try_insert(3..=100_000).unwrap();
+/// // now `tree` contains two intervals: [1..=1] and [3..=100_000]
+/// set.extend(3..=100_000);
+/// // extend complexity is O(n), where n is amount of elements in range.
+/// assert_eq!(set, tree.points_iter().collect());
+///
+/// // We can remove points from 1 to 99_000 not inclusive only by one remove operation.
+/// // `try` is only for range check, that it has start <= end.
+/// tree.try_remove(1..99_000).unwrap();
+/// // now `tree` contains two intervals: [99_000..=100_000]
+/// (1..99_000).for_each(|i| { set.remove(&i); });
+/// // remove complexity for set is O(n*log(m)),
+/// // where `n` is amount of elements in range and `m` size of set.
+/// assert_eq!(set, tree.points_iter().collect());
+///
+/// // Can insert or remove all possible points just by one operation:
+/// tree.insert(..);
+/// tree.remove(..);
+///
+/// // Iterate over voids (intervals between intervals in tree):
+/// tree.try_insert(1..=3).unwrap();
+/// tree.try_insert(5..=7).unwrap();
+/// let voids = tree.voids(..).map(RangeInclusive::from).collect::<Vec<_>>();
+/// assert_eq!(voids, vec![i32::MIN..=0, 4..=4, 8..=i32::MAX]);
+///
+/// // AndNot iterator: iterate over intervals from `tree` which are not in `other_tree`.
+/// let other_tree: IntervalsTree<i32> = [3, 4, 5, 7, 8, 9].iter().collect();
+/// let and_not: Vec<_> = tree.and_not_iter(&other_tree).map(RangeInclusive::from).collect();
+/// assert_eq!(and_not, vec![1..=2, 6..=6]);
+/// ```
+///
+/// # Possible panic cases.
+/// Using `IntervalsTree` for type `T: Numerated` cannot cause panics,
+/// if implementation `Numerated`, `Copy`, `Ord`, `Eq` are correct for `T`.
+/// In other cases `IntervalsTree` does not guarantees execution without panics.
+#[derive(Clone, PartialEq, Eq, TypeInfo, Encode, Decode)]
+pub struct IntervalsTree<T> {
+    inner: BTreeMap<T, T>,
+}
+
+impl<T: Copy> IntervalsTree<T> {
+    pub const fn new() -> Self {
+        Self {
+            inner: BTreeMap::new(),
+        }
+    }
+    pub fn intervals_amount(&self) -> usize {
+        self.inner.len()
+    }
+    pub fn end(&self) -> Option<T> {
+        self.inner.iter().next_back().map(|(_, &e)| e)
+    }
+
+    pub fn start(&self) -> Option<T> {
+        self.inner.iter().next().map(|(&s, _)| s)
+    }
+}
+
+impl<T: Copy> Default for IntervalsTree<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Debug> Debug for IntervalsTree<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{:?}",
+            self.inner.iter().map(|(s, e)| s..=e).collect::<Vec<_>>()
+        )
+    }
+}
+
+impl<T: Numerated> IntervalsTree<T> {
+    fn into_start_end<I: Into<Interval<T>>>(interval: I) -> Option<(T, T)> {
+        Into::<Interval<T>>::into(interval).into_inner()
+    }
+
+    /// Returns iterator over all intervals in tree.
+    pub fn iter(&self) -> impl Iterator<Item = NotEmptyInterval<T>> + '_ {
```

Comment:
```rust
impl<T> IntoIterator<Item = NotEmptyInterval<T>> for IntervalsTree<T> {}
```

Reply:
unfortunatelly it's impossible to use
```
IntoIter = impl Iterator<Item = NonEmptyInterval<T>> + '_
```
because it's unstable in rust, so I prefer to leave this as is